data_structures/src/chain.rs

```diff
@@ -56,18 +56,93 @@ pub struct CheckpointBeacon {
     /// The serial number for an epoch
     pub checkpoint: Epoch,
     /// The 256-bit hash of the previous block header
-    pub hash_prev_block: Vec<u8>,
+    pub hash_prev_block: Hash,
 }
 
 impl Default for CheckpointBeacon {
     fn default() -> CheckpointBeacon {
         CheckpointBeacon {
-            checkpoint: Epoch(0),
-            hash_prev_block: vec![0; 32],
+            checkpoint: 0,
+            hash_prev_block: Hash::SHA256([0; 32]),
         }
     }
 }
 
 /// Epoch id (starting from 0)
-#[derive(Copy, Clone, Debug, PartialEq, PartialOrd, Ord, Eq, Serialize, Deserialize)]
-pub struct Epoch(pub u64);
+pub type Epoch = u32;
+
+/// Block data structure
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct Block {
+    /// The header of the block
+    pub header: BlockHeaderWithProof,
+    /// The total number of transactions of the block
+    pub txn_count: u32,
+    /// A non-empty list of transactions
+    pub txns: Vec<Transaction>,
+}
+
+/// Block header structure
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct BlockHeader {
+    /// The block version number indicating the block validation rules
+    pub version: u32,
+    /// A checkpoint beacon for the epoch that this block is closing
+    pub beacon: CheckpointBeacon,
+    /// A 256-bit hash based on all of the transactions committed to this block
+    pub hash_merkle_root: Hash,
+}
+
+/// Block header structure with proof of existence
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct BlockHeaderWithProof {
```

Comment:
Instead of copying the fields from `BlockHeader`, I would define `BlockHeaderWithProof` as:

```rust
pub struct BlockHeaderWithProof {
    pub block_header: BlockHeader,
    pub proof: LeadershipProof,
}
```